docs/defining-data-models.md

```diff
@@ -0,0 +1,126 @@
+
+In a sentence, a `DataModel` is a set of `attributes`, and a `Description` of a service/resource is an assignment of those attributes.
+
+All you need to specify data models and descriptions (that is, instances of the data model) can be found in the `aea.helpers.search` module.
+
+
+## Attributes
+
+At the lowest level of our data model language, we have the `Attribute`.
+An attribute is an abstract definition of a property.
+
+It is identified by a `name`, that must be unique in a given data model (that is, we can't have two attributes that share the same name).
+
+Every attribute has a `type`, that specifies the domain of the property, that is, the possible values that the attribute can assume. At the moment, we support five types of attributes:
+
+* strings
+* integers
+* booleans
+* floats
+* locations, i.e. pairs of (latitude, longitude)
+
+An attribute can be `optional`, in the sense that instantiation of the attribute is not mandatory by the instances of the data model.
+
+Finally, every attribute might have a `description` that explains the purpose of the attribute.
+
+**Example**: suppose we have a bookshop, and we want to describe the books we sell. Presumably, we would like to include: the following properties of our books:
+
+* The `title`
+* The `author`
+* The `genre` (e.g. science fiction, horror)
+* The `year of publication`
+* The `average rating` (average of the ratings between 0 and 5)
+* The `ISBN` code
+* If it can be sold as an e-book.
+
+For each of this fields, we can define an attribute by using `Attribute`:
+
+``` python
+from aea.helpers.search import Attribute, Location
+attr_title   = Attribute("title", str, True, "The title of the book.")
+attr_author  = Attribute("author", str, True, "The author of the book.")
+attr_genre   = Attribute("genre", str, True, "The genre of the book.")
+attr_year    = Attribute("year", int, True, "The year of publication of the book.")
+attr_avg_rat = Attribute("average_rating",  float,    False, "The average rating of the book.")
+attr_isbn    = Attribute("ISBN", str, True, "The ISBN.")
+attr_ebook   = Attribute("ebook_available", bool, False, "If the book can be sold as an e-book.")
+attr_bookshop = Attribute("bookshop_pos", Location, False, "The location of the bookshop where you can find the book")
+```
+Let's focus on the parameters of the `Attribute` constructor:
+
+1. the first one is the name of the attribute. It is needed to instantiate a data model and to define queries over it.
+2. the second one is the type of the attribute. It specifies the domain of the possible values the attribute can assume.
+   E.g. the attribute ``year`` can only be an integer, whereas the ``average_rating`` can only be a floating-point number.
+   The supported types are: ``str``, ``int``, ``bool``, ``float`` and `Location`.
```

Comment:
Double backticks still work, but having single backticks looks better in Markdown